core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -803,6 +804,22 @@ trait EntityComponent {
         entityAttributeShardQuery(workspaceContext).deleteAttributes(workspaceContext, entityType, attributesNames)
     }
 
+    def doesAttributeNameAlreadyExist(workspaceContext: Workspace,
+                                      entityType: String,
+                                      newAttributeName: AttributeName): ReadAction[Option[Boolean]] = {
+      uniqueResult(entityAttributeShardQuery(workspaceContext).AttributeColumnQueries.doesAttributeExist(workspaceContext, entityType, newAttributeName))
+    }
+
+    def renameAttribute(workspaceContext: Workspace,
+                         entityType: String,
+                         oldAttributeName: AttributeName,
+                         newAttributeName: AttributeName): ReadWriteAction[Int] = {
+      for {
+        numRowsRenamed <- entityAttributeShardQuery(workspaceContext).AttributeColumnQueries.renameAttribute(workspaceContext, entityType, oldAttributeName, newAttributeName)
+        _ <- workspaceQuery.updateLastModified(workspaceContext.workspaceIdAsUUID)
+      } yield numRowsRenamed
+    }
```

Comment:
how about putting these inside `AttributeComponent` instead of `EntityComponent`? These feel like they should be parallel to `deleteAttributes()`:
https://github.com/broadinstitute/rawls/blob/4247d58764d24e2b30acc5350d7fd3e67002b399/core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala#L401-L403

Reply:
That makes sense! I was missing the fact that I should invalidate the statistics cache from EntityComponent, while keeping all the attribute logic in AttirbuteComponent-

Reply:
I see your point about extra boilerplate. The `renameAttribute` wrapper adds value - in that it updates the workspace's last-modified - but the `doesAttributeNameAlreadyExist` wrapper doesn't add anything. I'd be fine with deleting that.

There's no particular reason to only update the workspace's last-modified from EntityComponent as opposed to AttirbuteComponent.

In this PR as-is, we have the wrapper functions in BOTH EntityComponent and AttirbuteComponent, and the renameAttribute() implementation is different between the two; we definitely need to choose one. My suggestions is still to put the full renameAttribute() in AttributeComponent, and either put doesAttributeNameAlreadyExist() in AttributeComponent or delete it entirely. Maybe this needs a f2f chat, I'm not 100% sure I understand all of your reasoning.

Reply:
Ah, I see where we're disconnected. You initially referenced the "deleteAttributes" method, which consists of a "deleteAttributes" wrapper function in the EntityComponent that invalidates the EntityStatistics cache, and has another wrapper in the AttributeComponent, which doesn't add logic (the excerpt you have linked is the wrapper in AttributeComponent which doesn't add any logic). 

I was a bit confused about the need for the additional wrapper, but I followed the "deleteAttributes" pattern and updated the PR so each function had 2 wrappers, one in EntityComponent, and one in AttributeComponent. 

I added the comment pushing back because I wasn't feeling good about all the empty logic wrappers in the AttributeComponent. I'm a big fan of doing away with the wrappers, except the initial 2 wrappers from the first revision of the PR. This will be inconsistent with the DeleteAttributes pattern, which is invalidating the EntityStatistics cache in the EntityComponent, but I'll just move that over.

Happy to talk more about this face to face-